arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Either.kt

```diff
@@ -821,12 +821,54 @@ public sealed class Either<out A, out B> {
    * Example:
    * ```
    * Right(12).mapLeft { "flower" } // Result: Right(12)
-   * Left(12).mapLeft { "flower" }  // Result: Left("flower)
+   * Left(12).mapLeft { "flower" }  // Result: Left("flower")
    * ```
    */
   public inline fun <C> mapLeft(f: (A) -> C): Either<C, B> =
     fold({ Left(f(it)) }, { Right(it) })
 
+  /**
+   * The given function is applied as a fire and forget effect
+   * if this is a `Left`.
+   * When applied the result is ignored and the original
+   * Either value is returned
+   *
+   * Example:
+   * ```
+   * Right(12).tapLeft { println("flower") } // Result: Right(12)
+   * Left(12).tapLeft { println("flower") }  // Result: prints "flower" and returns: Left(12)
+   * ```
+   */
+  public inline fun tapLeft(f: (A) -> Unit): Either<A, B> =
+    when (this) {
+      is Left -> {
+        f(this.value)
+        this
+      }
+      is Right -> this
+    }
```

Comment:
We can also express this with a scoping function:
```suggestion
  public inline fun tapLeft(f: (A) -> Unit): Either<A, B> = apply {
    if(this is Left) {
       f(value)
    }
  }
```
Is it easier to read? 🤔

Reply:
If this feels better it would apply to the other taps 🙂 Otherwise happy to be ignored 😅 #notablocker

Reply:
@pablisco apply it's fine since I suppose the lambda has no cost, and its body is inlined since it's flagged as invoked once in its contract. I speculate it does not need to capture the state for the lambda. I'd be curious to see the resulting bytecode because the `when` statement I wrote has no allocations and most likely results in a table-switch indexed lookup.
In whatever case, It's not important to me. If you want to suggest some changes to the impl in the same branch, we can apply them before merging; I'm sure whatever users write in the lambda is gonna be more costly than any optimizations :)

Reply:
@raulraja Actually, now I think of it we could add the option to add effects to both sides of the bifunctor and make use of existing operations :slightly\_smiling\_face: 

```kotlin
  public inline fun onLeft(f: (A) -> Unit): Either<A, B> =
    onEither(left = f, right = {})

  public inline fun onRight(f: (B) -> Unit): Either<A, B> =
    onEither(left = {}, right = f)

  public inline fun onEither(left: (A) -> Unit, right: (B) -> Unit): Either<A, B> = 
    apply { fold(left, right) }
```

Reply:
that's probably allocating for each one of those lambdas? If that is the case, I'd rather have an alloc free impl that is less readable as we have no context as to how these functions will be used.

Reply:
That is probably right. One option may be to use a decoy for the empty lambda if we don't have something like this already. Need to check the byte code :)   
As I said, unless someone else sees this useful, we can explore it later on

Reply:
I think `apply` is confusing here since `apply` is typically for mutable instantiation logic, like builder syntax.
Which this definitely doesn't do.

I think the `when` code Raul has is the simplest 👍